extensions/amp-story-dev-tools/0.1/amp-story-dev-tools.js

```diff
@@ -32,21 +36,182 @@ const fontsToLoad = [
+            width="30"
+            height="30"
+            viewBox="0 0 30 30"
+            fill="none"
+          >
+            <circle cx="15" cy="15" r="15" fill="white"></circle>
+            <path
+              fill="#202125"
+              d="M19.5 9a1.5 1.5 0 011.5 1.5v9a1.5 1.5 0 01-1.5 1.5V9zM8.25 9a1.5 1.5 0 011.5-1.5h6.75A1.5 1.5 0 0118 9v12a1.5 1.5 0 01-1.5 1.5H9.75a1.5 1.5 0 01-1.5-1.5V9zM22.5 10.5c.621 0 1.125.504 1.125 1.125v6.75c0 .622-.504 1.125-1.125 1.125v-9z"
+            />
+          </svg>
+          <span class="i-amphtml-story-dev-tools-brand-text">
+            <span>WEB STORIES</span>
+            <span>DEV - TOOLS</span>
+          </span>
+        </span>
+        <div class="i-amphtml-story-dev-tools-tabs"></div>
+        <a
+          class="i-amphtml-story-dev-tools-button i-amphtml-story-dev-tools-close"
+        >
+          <span>OPEN STORY</span>
+          <svg
+            xmlns="http://www.w3.org/2000/svg"
+            width="14"
+            height="14"
+            viewBox="0 0 14 14"
+            fill="none"
+          >
+            <path
+              d="M9.9165 4.08333L9.094 4.90583L10.599 6.41667H4.6665V7.58333H10.599L9.094 9.08833L9.9165 9.91667L12.8332 7L9.9165 4.08333ZM2.33317 2.91667H6.99984V1.75H2.33317C1.6915 1.75 1.1665 2.275 1.1665 2.91667V11.0833C1.1665 11.725 1.6915 12.25 2.33317 12.25H6.99984V11.0833H2.33317V2.91667Z"
+              fill="black"
+            />
+          </svg>
+        </a>
+      </div>
+    </div>
+  `;
+};
+
+/**
+ * Updates the hashString with the dictionary<string, string> passed in
+ * @public
+ * @param {!Object} updates
+ * @param {!Window} win
+ */
+export function updateHash(updates, win) {
+  let queryHash = parseQueryString(win.location.hash);
+  queryHash = Object.assign(queryHash, updates);
+  win.location.hash = Object.entries(queryHash)
+    .filter((e) => e[1] != undefined)
+    .map((e) => e[0] + '=' + e[1])
+    .join('&');
+}
+
+/** @enum {string} */
+const DevToolsTab = {
+  PREVIEW: 'Preview',
+  PAGE_EXPERIENCE: 'Page Experience',
+  LOGS: 'Logs',
+};
+
 export class AmpStoryDevTools extends AMP.BaseElement {
   /** @param {!AmpElement} element */
   constructor(element) {
     super(element);
+
+    const hashParams = parseQueryString(this.win.location.hash);
+
+    this.win.document.title = `Story Dev-Tools (${this.win.document.title})`;
+
+    // TODO: Remove support for url queryParam when widely available.
+    /** @private {string} */
+    this.storyUrl_ = hashParams['url'] || this.win.location.href.split('#')[0];
+
+    /** @private {!DevToolsTab} get URL param for tab (eg: #tab=page-experience) or default to PREVIEW*/
+    this.currentTab_ =
+      Object.values(DevToolsTab).find(
+        (e) => e.toLowerCase().replace(' ', '-') === hashParams['tab']
+      ) || DevToolsTab.PREVIEW;
+
+    /** @private {!Object} maps tabs to contents */
+    this.tabContents_ = Object.values(DevToolsTab).reduce(
```

Comment:
We should not do any work in the `constructor` such as creating DOM, even if not attached, please move this to `buildCallback`